pkg/controller/elasticsearch/driver/upgrade_deletion_strategy.go

```diff
@@ -0,0 +1,263 @@
+		toUpdate:         podsToUpgrade,
+		esState:          state,
+	}
+}
+
+// SortFunction is the default sort function, masters have lower priority as
+// we want to update the data nodes first.
+// If 2 Pods are of the same type then use the reverse ordinal order.
+// TODO: Add some priority to unhealthy (bootlooping) Pods
+func sortCandidates(allPods []corev1.Pod) {
+	sort.Slice(allPods, func(i, j int) bool {
+		pod1 := allPods[i]
+		pod2 := allPods[j]
+		if (label.IsMasterNode(pod1) && label.IsMasterNode(pod2)) ||
+			(!label.IsMasterNode(pod1) && !label.IsMasterNode(pod2)) { // same type, use the reverse name function
+			ssetName1, ord1, err := sset.StatefulSetName(pod1.Name)
+			if err != nil {
+				return false
+			}
+			ssetName2, ord2, err := sset.StatefulSetName(pod2.Name)
+			if err != nil {
+				return false
+			}
+			if strings.Compare(ssetName1, ssetName2) == 0 {
+				// same name, compare ordinal, higher first
+				return ord1 > ord2
+			}
+			return strings.Compare(ssetName1, ssetName2) == -1
+		}
+		if label.IsMasterNode(pod1) && !label.IsMasterNode(pod2) {
+			// pod2 has higher priority since it is a data node
+			return false
+		}
+		return true
+	})
+}
+
+var predicates = [...]Predicate{
+	{
+		// If MaxUnavailable is reached, allow for an unhealthy Pod to be deleted.
+		// This is to prevent a situation where MaxUnavailable is reached and we
+		// can't make some progress even if the user has updated the spec.
+		name: "do_not_restart_healthy_node_if_MaxUnavailable_reached",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			if maxUnavailableReached && k8s.IsPodReady(candidate) {
+				return false, nil
+			}
+			return true, nil
+		},
+	},
+	{
+		name: "skip_unknown_or_long_terminating_pods",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			if candidate.DeletionTimestamp != nil && candidate.Status.Reason == NodeUnreachablePodReason {
+				// kubelet is unresponsive, Unknown Pod, do not try to delete it
+				return false, nil
+			}
+			if candidate.DeletionTimestamp != nil && time.Now().After(candidate.DeletionTimestamp.Add(TerminatingPodTimeout)) {
+				return false, nil
+			}
+			return true, nil
+		},
+	},
+	{
+		// In Yellow or Red status only allow unhealthy Pods to be restarted.
+		// This is intended to unlock some situations where the cluster is not green and
+		// a Pod has to be restarted a second time.
+		name: "do_not_restart_healthy_node_if_not_green",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			green, err := context.esState.GreenHealth()
+			if err != nil {
+				return false, err
+			}
+			if green {
+				return true, nil
+			}
+			if !k8s.IsPodReady(candidate) {
```

Comment:
Looks like we have two different notions of healthy in this predicate: `ctx.healthyPods` which include ES cluster membership and  this which is just pod readiness. Is there a reason not to use `ctx.healthyPods` here?